sidechain/consensus/common/src/test/mocks/block_import_queue_worker_mock.rs

```diff
@@ -104,27 +146,135 @@ mod tests {
+				if i % 2 == 0 && i != 1 { // 1 is not even want all odds to have 2 of the same block_number
+					let header = <BlockQueueHeaderBuilder<u64, H256>>::build_queue_header(i, parent_header.hash());
+					queue.push_back(SidechainBlockBuilder::default().with_header(header).build());
+				} else {
+					// build a Queue with 2 headers which are of the same block_number
+					let mut headers = vec![
+						<BlockQueueHeaderBuilder<u64, H256>>::build_queue_header(i, parent_header.hash()),
+						<BlockQueueHeaderBuilder<u64, H256>>::build_queue_header(i, parent_header.hash())
+					];
+					headers.iter().for_each(|header| {
+						queue.push_back(SidechainBlockBuilder::default().with_header(*header).build());
+					});
+				}
+			}
+			queue
+		});
+
+		// queue -> [0, 1, 1, 2, 3, 3, 4, 5, 5, 6, 7, 7]
+		assert_eq!(queue.len(), 12);
+
+		// Store all block_headers in db
+		let mut db = <HeaderDb<H256, Header>>::from(
+			queue.iter().map(|block| (block.hash(), *block.header())).collect::<Vec<_>>().as_slice()
+		);
+
+		// Import into forktree
+		let is_descendent_of = 
+			<IsDescendentOfBuilder<H256, HeaderDb<H256, Header>, TestError>>::build_is_descendent_of(None, &db);
+		let mut tree = <ForkTree<H256, u64, ()>>::new();
+		queue.iter().for_each(|block| {
+			let _ = tree.import(block.header.hash(), block.header.block_number(), (), &is_descendent_of).unwrap();
+		});
+
+		// We have a tree which looks like the following
+		//                      - (H5, B3)..
+		//                     /
+		//		 	 - (H3, B2)
+		//			/          \
+		//   	 - (H1, B1)     - (H4, B3)..
+		//  	/
+		//	   /	 
+		// (H0, B0)   
+		//     \
+		//  	\	
+		//		 - (H2, B1)..
 		//
-		//   - H2..
-		//  /
-		// H1..   - H4..
-		//  \   /
-		//   - H3..
-		//      \
-		//       - H5..
 		//
-		todo!();
-		println!("Process Sequential Queue with Forks")
+
+		// H0 is the first root
+		assert_eq!(
+			tree.roots().map(|(h, n, _)| (*h, *n)).collect::<Vec<_>>(),
+			vec![(queue.front().unwrap().header.hash(), 0)]
+		);
+
+		// Now if we finalize H0 we should see 2 roots H1 and H2
+		tree.finalize_root(&queue.front().unwrap().header.hash()).unwrap();
+		let _ = queue.pop_front();
+		assert_eq!(
+			tree.roots().map(|(h, n, _)| (*h, *n)).collect::<Vec<_>>(),
+			vec![(queue[1].header.hash(), 1), (queue[0].header.hash(), 1)]
+		);
+
+		// If we finalize (H1, B1) then we should see one roots (H3, B2)
+		let _ = queue.pop_front(); // remove (H1, B1)
+		tree.finalize_root(&queue.front().unwrap().header.hash()).unwrap();
+		let _ = queue.pop_front(); // remove (H2, B1)
+		assert_eq!(
+			tree.roots().map(|(h, n, _)| (*h, *n)).collect::<Vec<_>>(),
+			vec![(queue[0].header.hash(), 2)]
+		);
+
+		// If we finalize (H3, B2) we should see two roots (H4, B3), (H5, B3)
+		tree.finalize_root(&queue.front().unwrap().header.hash()).unwrap();
+		let _ = queue.pop_front(); // remove (H3, B2)
+		assert_eq!(
+			tree.roots().map(|(h, n, _)| (*h, *n)).collect::<Vec<_>>(),
```

Comment:
Maybe we want to add a helper method that enables

```rust
// iterate over hash and number of the roots.
tree.roots().hash_and_number()
```

This will make the code much more readable.